### PR TITLE
1004 OAuth2 client credentials flow

### LINKED ISSUE #1004: OAuth2 Client Credentials Flow
### Is your feature request related to a specific problem?

We currently only have the OAuth2 Authentication Code flow (grant) implemented. Some APIs we are integrating will require the Client Credentials Flow. Although the way the token request is built for the client credentials flow is the same as the authentication code flow, the main difference is that the client credentials flow does not require an authorization step. This means we don't need to present the user with an authorization link or validate the same configuration as the auth code flow.

- Auth code flow/grant (https://www.oauth.com/oauth2-servers/server-side-apps/authorization-code/)
- Client credentials flow/grant (https://www.oauth.com/oauth2-servers/access-tokens/client-credentials/)

**Impact:**
Affects the auth0 connector. We currently don't support this oauth flow. This flow is different enough to warrant it's own strategy. Mostly just need to reorganize the code to allow for this new flow.

There may be an interaction with this ticket https://github.com/ethyca/fidesops/issues/562 but it should be easy enough to resolve any conflicts.

### Describe the solution you'd like

- Refactor the common oauth2 strategy logic into a base class
- Make the authentication code flow a subclass of the base oauth2 strategy (`oauth2_authentication_code`)
- Make the client credentials flow a subclass of the base oauth2 strategy (`oauth2_client_credentials`)

### Describe alternatives you've considered, if any

I considered keeping on the oauth2 strategy and branching when necessary when the logic differed between auth code and client credentials flow but I decided it would be best to move the shared oauth2 logic to a base class and have these two flows just be subclasses. This way the logic is easier to follow between flows.

### PR DESCRIPTION
**Pull request by galvana**
# Purpose
Adding the client credential OAuth2 flow. We previously only supported the authentication code flow but we need to be able to support additional flows for different SaaS connectors.

# Changes
- Renamed the existing `oauth2` strategy to `oauth2_authentication_code` and added `oauth2_client_credentials`
- Moved shared OAuth2 functionality to a base class which is extended for each OAuth2 flow
- Added unit tests for Client Credentials flow
- Updated documentation to include the new flow

# Checklist
- [x] Update [`CHANGELOG.md`](https://github.com/ethyca/fidesops/blob/main/CHANGELOG.md) file
  - [x] Merge in main so the most recent `CHANGELOG.md` file is being appended to
  - [x] Add description within the `Unreleased` section in an appropriate category. Add a new category from the list at the top of the file if the needed one isn't already there.
  - [x] Add a link to this PR at the end of the description with the PR number as the text. example: [#1](https://github.com/ethyca/fidesops/pull/1)
- [x] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md).
- If docs updated (select one):
  - [x] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
- [x] Good unit test/integration test coverage
- [x] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Fixes #1004

### CODE REVIEW

File: src/fidesops/ops/api/v1/endpoints/oauth_endpoints.py [hunk 1]
```diff
@@ -214,9 +214,10 @@ def oauth_callback(code: str, state: str, db: Session = Depends(get_db)) -> None
         authentication = (
             connection_config.get_saas_config().client_config.authentication  # type: ignore
         )
-        auth_strategy: OAuth2AuthenticationStrategy = get_strategy(  # type: ignore
+        auth_strategy: OAuth2AuthorizationCodeAuthenticationStrategy = get_strategy(  # type: ignore
             authentication.strategy, authentication.configuration  # type: ignore
         )
-        auth_strategy.get_access_token(db, code, connection_config)
+        connection_config.secrets = {**connection_config.secrets, "code": code}  # type: ignore
```
Comment (galvana) [inline src/fidesops/ops/api/v1/endpoints/oauth_endpoints.py:220]: Adding the code outside of the `get_access_token` to make the function reusable between the different OAuth2 flows.

File: src/fidesops/ops/service/authentication/authentication_strategy_oauth2_base.py [hunk 2]
```diff
@@ -216,7 +174,7 @@ def _validate_and_store_response(
         # This alternate way of specifying the expiration is handled
         # by the optional expires_in field of the OAuth2AuthenticationConfiguration
 
-        expires_in = response.get("expires_in") or self.expires_in
+        expires_in = self.expires_in or response.get("expires_in")
```
Comment (galvana) [inline src/fidesops/ops/service/authentication/authentication_strategy_oauth2_base.py:177]: Switching order to be able to override expiration even if one is provided in the token response.

File: src/fidesops/ops/api/v1/endpoints/saas_config_endpoints.py [hunk 3]
```diff
@@ -259,7 +262,7 @@ def authorize_connection(
     authentication = connection_config.get_saas_config().client_config.authentication  # type: ignore
 
     try:
-        auth_strategy: OAuth2AuthenticationStrategy = get_strategy(
+        auth_strategy: OAuth2AuthorizationCodeAuthenticationStrategy = get_strategy(
```
Comment (adamsachs) [inline src/fidesops/ops/api/v1/endpoints/saas_config_endpoints.py:265]: i smell a merge conflict with https://github.com/ethyca/fidesops/pull/1163 😅

File: tests/ops/service/authentication/test_authentication_strategy_oauth2_client_credentials.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+    def test_get_access_token(
+        self,
+        mock_send: Mock,
+        mock_connection_config_update: Mock,
+        mock_time: Mock,
+        db: Session,
+        oauth2_client_credentials_connection_config,
+        oauth2_client_credentials_configuration,
+    ):
+        # cast some time magic
```
Comment (adamsachs) [inline tests/ops/service/authentication/test_authentication_strategy_oauth2_client_credentials.py:323]: who said time travel wasn't possible

File: tests/ops/service/authentication/test_authentication_strategy_oauth2_client_credentials.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+                    "client_id": "client",
+                    "client_secret": "secret",
+                    "access_token": "new_access",
+                    "refresh_token": "new_refresh",
+                    "expires_at": int(datetime.utcnow().timestamp()) + expires_in,
+                }
+            },
+        )
+
+    # make sure we can use the expires_in value in the config if no expires_in is provided
```
Comment (adamsachs) [inline tests/ops/service/authentication/test_authentication_strategy_oauth2_client_credentials.py:354]: nit: should this comment be moved to the docstring for the test method?

**Review by adamsachs — APPROVED** [review]
this looks really good to me - nice job abstracting out the common components. i have only some very minor comments - please let me know what you think, but i don't consider them blocking.

also, just noting that we'll need to resolve merge conflicts in https://github.com/ethyca/fidesops/pull/1163 once this is merged.

**Review by adamsachs — CHANGES_REQUESTED** [review]
sorry, i spoke a bit too soon on my previous review -- i shouldn't have approved yet because you're missing a `CHANGELOG.md` update. i actually think this change should be logged as under `Breaking Changes` because technically it would break any existing configs that have the old `oauth2` auth strategy name.

File: src/fidesops/ops/service/authentication/authentication_strategy_oauth2_base.py [hunk 6]
```diff
@@ -28,80 +22,34 @@
...
-    strategy_name = "oauth2"
-
-    def __init__(self, configuration: OAuth2AuthenticationConfiguration):
+    def __init__(self, configuration: OAuth2BaseConfiguration) -> None:
         self.expires_in = configuration.expires_in
-        self.authorization_request = configuration.authorization_request
         self.token_request = configuration.token_request
         self.refresh_request = configuration.refresh_request
 
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/ops/service/authentication/authentication_strategy_oauth2_base.py:36]: Nice catch, I deleted the re-definition

**Comment by galvana** [comment]
Updated the changelog with the breaking change and made the requested changes.

**Review by adamsachs — APPROVED** [review]
all looks good to me, thanks for addressing the minor comments!

**Pull request merged**